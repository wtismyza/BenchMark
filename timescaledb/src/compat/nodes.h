/*
 * This file and its contents are licensed under the Apache License 2.0.
 * Please see the included NOTICE for copyright information and
 * LICENSE-APACHE for a copy of the license.
 */
#ifndef TIMESCALEDB_COMPAT_NODES_H
#define TIMESCALEDB_COMPAT_NODES_H

#if PG12_LT

#define T_SubscriptingRef T_ArrayRef
typedef ArrayRef SubscriptingRef;

#endif

#endif /* TIMESCALEDB_COMPAT_NODES_H */
