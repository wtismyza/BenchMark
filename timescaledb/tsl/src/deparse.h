/*
 * This file and its contents are licensed under the Timescale License.
 * Please see the included NOTICE for copyright information and
 * LICENSE-TIMESCALE for a copy of the license.
 */
#ifndef TIMESCALEDB_DEPARSE_H
#define TIMESCALEDB_DEPARSE_H

#include <postgres.h>
#include <nodes/pg_list.h>

typedef struct TableInfo
{
	Oid relid;
	List *constraints;
	List *indexes;
	List *triggers;
	List *rules;
} TableInfo;

typedef struct TableDef
{
	const char *schema_cmd;
	const char *create_cmd;
	List *constraint_cmds;
	List *index_cmds;
	List *trigger_cmds;
	List *rule_cmds;
} TableDef;

typedef struct DeparsedHypertableCommands
{
	const char *table_create_command;
	List *dimension_add_commands;
	List *grant_commands;
} DeparsedHypertableCommands;

typedef struct Hypertable Hypertable;

TableInfo *deparse_create_table_info(Oid relid);
TableDef *deparse_get_tabledef(TableInfo *table_info);
List *deparse_get_tabledef_commands(Oid relid);
List *deparse_get_tabledef_commands_from_tabledef(TableDef *table_def);
const char *deparse_get_tabledef_commands_concat(Oid relid);

DeparsedHypertableCommands *deparse_get_distributed_hypertable_create_command(Hypertable *ht);

const char *deparse_func_call(FunctionCallInfo finfo);
const char *deparse_oid_function_call_coll(Oid funcid, Oid collation, unsigned int num_args, ...);

#endif
