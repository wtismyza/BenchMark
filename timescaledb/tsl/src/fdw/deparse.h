/*
 * This file and its contents are licensed under the Timescale License.
 * Please see the included NOTICE for copyright information and
 * LICENSE-TIMESCALE for a copy of the license.
 */
#ifndef TIMESCALEDB_TSL_FDW_DEPARSE_H
#define TIMESCALEDB_TSL_FDW_DEPARSE_H

#include <postgres.h>
#include "data_node_chunk_assignment.h"

typedef struct DeparsedInsertStmt
{
	const char *target; /* INSERT INTO (...) */
	unsigned int num_target_attrs;
	const char *target_attrs;
	bool do_nothing;
	const char *returning;
	List *retrieved_attrs;
} DeparsedInsertStmt;

extern void deparse_insert_stmt(DeparsedInsertStmt *stmt, RangeTblEntry *rte, Index rtindex,
								Relation rel, List *target_attrs, bool do_nothing,
								List *returning_list);

extern List *deparsed_insert_stmt_to_list(DeparsedInsertStmt *stmt);
extern void deparsed_insert_stmt_from_list(DeparsedInsertStmt *stmt, List *list_stmt);

extern const char *deparsed_insert_stmt_get_sql(DeparsedInsertStmt *stmt, int64 num_rows);
extern const char *deparsed_insert_stmt_get_sql_explain(DeparsedInsertStmt *stmt, int64 num_rows);

extern void deparseInsertSql(StringInfo buf, RangeTblEntry *rte, Index rtindex, Relation rel,
							 List *targetAttrs, int64 num_rows, bool doNothing, List *returningList,
							 List **retrieved_attrs);

extern void deparseUpdateSql(StringInfo buf, RangeTblEntry *rte, Index rtindex, Relation rel,
							 List *targetAttrs, List *returningList, List **retrieved_attrs);

extern void deparseDeleteSql(StringInfo buf, RangeTblEntry *rte, Index rtindex, Relation rel,
							 List *returningList, List **retrieved_attrs);

extern bool is_foreign_expr(PlannerInfo *root, RelOptInfo *baserel, Expr *expr);

extern void classify_conditions(PlannerInfo *root, RelOptInfo *baserel, List *input_conds,
								List **remote_conds, List **local_conds);

extern List *build_tlist_to_deparse(RelOptInfo *foreignrel);

extern void deparseSelectStmtForRel(StringInfo buf, PlannerInfo *root, RelOptInfo *rel, List *tlist,
									List *remote_conds, List *pathkeys, bool is_subquery,
									List **retrieved_attrs, List **params_list,
									DataNodeChunkAssignment *swa, List **current_time_idx);

extern const char *get_jointype_name(JoinType jointype);
extern void deparseStringLiteral(StringInfo buf, const char *val);
extern void deparseAnalyzeSizeSql(StringInfo buf, Relation rel);
extern void deparseAnalyzeSql(StringInfo buf, Relation rel, List **retrieved_attrs);

#endif /* TIMESCALEDB_TSL_FDW_DEPARSE_H */
