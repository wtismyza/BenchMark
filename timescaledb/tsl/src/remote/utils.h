/*
 * This file and its contents are licensed under the Timescale License.
 * Please see the included NOTICE for copyright information and
 * LICENSE-TIMESCALE for a copy of the license.
 */
#ifndef TIMESCALEDB_TSL_REMOTE_UTILS_H
#define TIMESCALEDB_TSL_REMOTE_UTILS_H

#include <postgres.h>
#include <foreign/foreign.h>

extern int set_transmission_modes(void);
extern void reset_transmission_modes(int nestlevel);

#endif /* TIMESCALEDB_TSL_REMOTE_UTILS_H */
