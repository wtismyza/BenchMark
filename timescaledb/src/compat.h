/*
 * This file and its contents are licensed under the Apache License 2.0.
 * Please see the included NOTICE for copyright information and
 * LICENSE-APACHE for a copy of the license.
 */
#ifndef TIMESCALEDB_COMPAT_H
#define TIMESCALEDB_COMPAT_H

#include <postgres.h>
#include <commands/explain.h>
#include <commands/trigger.h>
#include <executor/executor.h>
#include <executor/tuptable.h>
#include <nodes/execnodes.h>
#include <nodes/nodes.h>
#include <pgstat.h>
#include <utils/lsyscache.h>
#include <utils/rel.h>

#include "export.h"

#define is_supported_pg_version_11(version) ((version >= 110000) && (version < 120000))
#define is_supported_pg_version_12(version) ((version >= 120000) && (version < 130000))

#define is_supported_pg_version(version)                                                           \
	(is_supported_pg_version_11(version) || is_supported_pg_version_12(version))

#define PG11 is_supported_pg_version_11(PG_VERSION_NUM)
#define PG12 is_supported_pg_version_12(PG_VERSION_NUM)

#define PG12_LT PG11
#define PG12_GE !(PG12_LT)

#if !(is_supported_pg_version(PG_VERSION_NUM))
#error "Unsupported PostgreSQL version"
#endif

/*
 * The following are compatibility functions for different versions of
 * PostgreSQL. Each compatibility function (or group) has its own logic for
 * which versions get different behavior and is separated from others by a
 * comment with its name and any clarifying notes about supported behavior. Each
 * compatibility define is separated out by function so that it is easier to see
 * what changed about its behavior, and at what version, but closely related
 * functions that changed at the same time may be grouped together into a single
 * block. Compatibility functions are organized in alphabetical order.
 *
 * Wherever reasonable, we try to achieve forwards compatibility so that we can
 * take advantage of features added in newer PG versions. This avoids some
 * future tech debt, though may not always be possible.
 *
 * We append "compat" to the name of the function or define if we change the behavior
 * of something that existed in a previous version. If we are merely backpatching
 * behavior from a later version to an earlier version and not changing the
 * behavior of the new version we simply adopt the new version's name.
 */

#if PG11
#define ExecInsertIndexTuplesCompat(slot, estate, no_dup_err, spec_conflict, arbiter_indexes)      \
	ExecInsertIndexTuples(slot,                                                                    \
						  &((slot)->tts_tuple->t_self),                                            \
						  estate,                                                                  \
						  no_dup_err,                                                              \
						  spec_conflict,                                                           \
						  arbiter_indexes)
#else
#define ExecInsertIndexTuplesCompat(slot, estate, no_dup_err, spec_conflict, arbiter_indexes)      \
	ExecInsertIndexTuples(slot, estate, no_dup_err, spec_conflict, arbiter_indexes);
#endif

/* ExecARInsertTriggers */
#if PG11
#define ExecARInsertTriggersCompat(estate, relinfo, slot, recheck_indexes, transition_capture)     \
	do                                                                                             \
	{                                                                                              \
		bool should_free;                                                                          \
		HeapTuple tuple = ExecFetchSlotHeapTuple(slot, true, &should_free);                        \
		ExecARInsertTriggers(estate, relinfo, tuple, recheck_indexes, transition_capture);         \
		if (should_free)                                                                           \
			heap_freetuple(tuple);                                                                 \
	} while (0);
#else
#define ExecARInsertTriggersCompat(estate,                                                         \
								   result_rel_info,                                                \
								   tuple,                                                          \
								   recheck_indexes,                                                \
								   transition_capture)                                             \
	ExecARInsertTriggers(estate, result_rel_info, tuple, recheck_indexes, transition_capture)
#endif

/* execute_attr_map_tuple */
#if PG11
#define execute_attr_map_tuple do_convert_tuple
#endif

#if PG11
#define TM_Result HTSU_Result

#define TM_Ok HeapTupleMayBeUpdated
#define TM_SelfModified HeapTupleSelfUpdated
#define TM_Updated HeapTupleUpdated
#define TM_BeingModified HeapTupleBeingUpdated
#define TM_WouldBlock HeapTupleWouldBlock
#define TM_Invisible HeapTupleInvisible

#define TM_FailureData HeapUpdateFailureData
#endif

#if PG11

#define TupleTableSlotOps void
#define TTSOpsVirtualP NULL
#define TTSOpsHeapTupleP NULL
#define TTSOpsMinimalTupleP NULL
#define TTSOpsBufferHeapTupleP NULL

#else

#define TTSOpsVirtualP (&TTSOpsVirtual)
#define TTSOpsHeapTupleP (&TTSOpsHeapTuple)
#define TTSOpsMinimalTupleP (&TTSOpsMinimalTuple)
#define TTSOpsBufferHeapTupleP (&TTSOpsBufferHeapTuple)

#endif

/*
 * ExecInitExtraTupleSlot & MakeTupleTableSlot
 *
 * PG11 introduced the ability to pass in a tupledesc and avoid having to
 * separately run ExecSetSlotDescriptor. Additionally, it added the ability to have a
 * fixed tupleDesc when making a TupleTableSlot, which can be useful during
 * execution for JITted operations. (See:
 * https://github.com/postgres/postgres/commit/ad7dbee368a7cd9e595d2a957be784326b08c943).
 * We adopt the PG11 conventions so that we can take advantage of JITing more easily in the future.
 */
#if PG11

#define ExecInitExtraTupleSlotCompat(estate, tupledesc, tts_ops)                                   \
	ExecInitExtraTupleSlot(estate, tupledesc)
#define MakeTupleTableSlotCompat(tupdesc, tts_ops) MakeTupleTableSlot(tupdesc)
#define MakeSingleTupleTableSlotCompat(tupdesc, tts_ops) MakeSingleTupleTableSlot(tupdesc)
#define ExecStoreHeapTupleCompat(tuple, slot, should_free)                                         \
	ExecStoreTuple(tuple, slot, InvalidBuffer, should_free)
#define ExecForceStoreHeapTupleCompat(tuple, slot, should_free)                                    \
	ExecStoreTuple(tuple, slot, InvalidBuffer, should_free)

#else /* PG12_GE */

#define ExecInitExtraTupleSlotCompat(estate, tupdesc, tts_ops)                                     \
	ExecInitExtraTupleSlot(estate, tupdesc, tts_ops)
#define MakeTupleTableSlotCompat(tupdesc, tts_ops) MakeTupleTableSlot(tupdesc, tts_ops)
#define MakeSingleTupleTableSlotCompat(tupdesc, tts_ops) MakeSingleTupleTableSlot(tupdesc, tts_ops)
#define ExecStoreHeapTupleCompat(tuple, slot, should_free)                                         \
	ExecStoreHeapTuple(tuple, slot, should_free)
#define ExecForceStoreHeapTupleCompat(tuple, slot, should_free)                                    \
	ExecForceStoreHeapTuple(tuple, slot, should_free)

#endif

/* fmgr
 * In a9c35cf postgres changed how it calls SQL functions so that the number of
 * argument-slots allocated is chosen dynamically, instead of being fixed. This
 * change was ABI-breaking, so we cannot backport this optimization, however,
 * we do backport the interface, so that all our code will be compatible with
 * new versions.
 */
#if PG11

/* unlike the pg12 version, this is just a wrapper for FunctionCallInfoData */
#define LOCAL_FCINFO(name, nargs)                                                                  \
	union                                                                                          \
	{                                                                                              \
		FunctionCallInfoData fcinfo;                                                               \
	} name##data;                                                                                  \
	FunctionCallInfo name = &name##data.fcinfo

#define SizeForFunctionCallInfo(nargs) sizeof(FunctionCallInfoData)

/* convenience macro to allocate FunctionCallInfoData on the heap */
#define HEAP_FCINFO(nargs) palloc(sizeof(FunctionCallInfoData))

/* getting arguments has a different API, so these macros unify the versions */
#define FC_ARG(fcinfo, n) ((fcinfo)->arg[(n)])
#define FC_NULL(fcinfo, n) ((fcinfo)->argnull[(n)])

#else

/* convenience macro to allocate FunctionCallInfoData on the heap */
#define HEAP_FCINFO(nargs) palloc(SizeForFunctionCallInfo(nargs))

/* getting arguments has a different API, so these macros unify the versions */
#define FC_ARG(fcinfo, n) ((fcinfo)->args[(n)].value)
#define FC_NULL(fcinfo, n) ((fcinfo)->args[(n)].isnull)

#endif

#define FC_FN_OID(fcinfo) ((fcinfo)->flinfo->fn_oid)

/* convenience setters */
#define FC_SET_ARG(fcinfo, n, val)                                                                 \
	do                                                                                             \
	{                                                                                              \
		short _n = (n);                                                                            \
		FunctionCallInfo _fcinfo = (fcinfo);                                                       \
		FC_ARG(_fcinfo, _n) = (val);                                                               \
		FC_NULL(_fcinfo, _n) = false;                                                              \
	} while (0)

#define FC_SET_NULL(fcinfo, n)                                                                     \
	do                                                                                             \
	{                                                                                              \
		short _n = (n);                                                                            \
		FunctionCallInfo _fcinfo = (fcinfo);                                                       \
		FC_ARG(_fcinfo, _n) = 0;                                                                   \
		FC_NULL(_fcinfo, _n) = true;                                                               \
	} while (0)

/*
 * In PG12 OID columns were removed changing all OID columns in the catalog to
 * be regular columns. This necessitates passing in the attnum of said column to
 * any function that wishes to access these columns. In earlier versions, this
 * parameter can be safely ignored.
 */
#if PG11
#define GetSysCacheOid1Compat(cacheid, oidcol, key1) GetSysCacheOid1(cacheid, key1)
#define GetSysCacheOid2Compat(cacheid, oidcol, key1, key2) GetSysCacheOid2(cacheid, key1, key2)
#define GetSysCacheOid3Compat(cacheid, oidcol, key1, key2, key3)                                   \
	GetSysCacheOid3(cacheid, key1, key2, key3)
#define GetSysCacheOid4Compat(cacheid, oidcol, key1, key2, key3, key4)                             \
	GetSysCacheOid4(cacheid, key1, key2, key3, key4)
#else
#define GetSysCacheOid1Compat(cacheid, oidcol, key1) GetSysCacheOid1(cacheid, oidcol, key1)
#define GetSysCacheOid2Compat(cacheid, oidcol, key1, key2)                                         \
	GetSysCacheOid2(cacheid, oidcol, key1, key2)
#define GetSysCacheOid3Compat(cacheid, oidcol, key1, key2, key3)                                   \
	GetSysCacheOid3(cacheid, oidcol, key1, key2, key3)
#define GetSysCacheOid4Compat(cacheid, oidcol, key1, key2, key3, key4)                             \
	GetSysCacheOid4(cacheid, oidcol, key1, key2, key3, key4)
#endif

/* RenameRelationInternal
 */
#if PG11
#define RenameRelationInternalCompat(relid, name, is_internal, is_index)                           \
	RenameRelationInternal(relid, name, is_internal)
#else
#define RenameRelationInternalCompat RenameRelationInternal
#endif

/* create this function for symmetry with pq_sendint32 */
#define pq_getmsgint32(buf) pq_getmsgint(buf, 4)

#if PG11
#define TUPLE_DESC_HAS_OIDS(desc) (desc)->tdhasoid
#else
#define TUPLE_DESC_HAS_OIDS(desc) false
#endif

/* Compatibility functions for table access method API introduced in PG12 */
#if PG11
#include "compat/tupconvert.h"
#include "compat/tuptable.h"
#include "compat/tableam.h"

#else
#define ts_tuptableslot_set_table_oid(slot, table_oid) (slot)->tts_tableOid = table_oid
#endif

#if PG11
#define ExecTypeFromTLCompat(tlist, hasoid) ExecTypeFromTL(tlist, hasoid)
#else
#define ExecTypeFromTLCompat(tlist, hasoid) ExecTypeFromTL(tlist)
#endif

#if PG11
#define create_append_path_compat(root,                                                            \
								  rel,                                                             \
								  subpaths,                                                        \
								  partial_subpaths,                                                \
								  pathkeys,                                                        \
								  required_outer,                                                  \
								  parallel_workers,                                                \
								  parallel_aware,                                                  \
								  partitioned_rels,                                                \
								  rows)                                                            \
	create_append_path(root,                                                                       \
					   rel,                                                                        \
					   subpaths,                                                                   \
					   partial_subpaths,                                                           \
					   required_outer,                                                             \
					   parallel_workers,                                                           \
					   parallel_aware,                                                             \
					   partitioned_rels,                                                           \
					   rows)

#else
#define create_append_path_compat create_append_path
#endif

/*
 * estimate_hashagg_tablesize is a static function in PG11 and earlier, so we map
 * to our own copy when it's not available.
 */
#if PG11
#define estimate_hashagg_tablesize(p, c, n) ts_estimate_hashagg_tablesize(p, c, n)
#endif

#include <commands/vacuum.h>
#include <commands/defrem.h>

static inline int
get_vacuum_options(const VacuumStmt *stmt)
{
#if PG12_GE
	/* In PG12, the vacuum options is a list of DefElems and require
	 * parsing. Here we only parse the options we might be interested in since
	 * PostgreSQL itself will parse the options fully when it executes the
	 * vacuum. */
	ListCell *lc;
	bool analyze = false;
	bool verbose = false;

	foreach (lc, stmt->options)
	{
		DefElem *opt = (DefElem *) lfirst(lc);

		/* Parse common options for VACUUM and ANALYZE */
		if (strcmp(opt->defname, "verbose") == 0)
			verbose = defGetBoolean(opt);
		/* Parse options available on VACUUM */
		else if (strcmp(opt->defname, "analyze") == 0)
			analyze = defGetBoolean(opt);
	}

	return (stmt->is_vacuumcmd ? VACOPT_VACUUM : VACOPT_ANALYZE) | (verbose ? VACOPT_VERBOSE : 0) |
		   (analyze ? VACOPT_ANALYZE : 0);
#else
	return stmt->options;
#endif
}

#endif /* TIMESCALEDB_COMPAT_H */
