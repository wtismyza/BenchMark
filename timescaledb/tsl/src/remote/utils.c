/*
 * This file and its contents are licensed under the Timescale License.
 * Please see the included NOTICE for copyright information and
 * LICENSE-TIMESCALE for a copy of the license.
 */

/*
 * This file contains source code that was copied and/or modified from
 * the PostgreSQL database, which is licensed under the open-source
 * PostgreSQL License. Please see the NOTICE at the top level
 * directory for a copy of the PostgreSQL License.
 */
#include <postgres.h>
#include <miscadmin.h>
#include <catalog/pg_type.h>
#include <access/reloptions.h>
#include <access/htup_details.h>
#include <utils/syscache.h>
#include <utils/guc.h>
#include <utils/builtins.h>

#include <compat.h>
#if PG12_GE
#include <utils/float.h>
#endif
#include "utils.h"
#include "compat.h"
#include "guc.h"

/*
 * Force assorted GUC parameters to settings that ensure that we'll output
 * data values in a form that is unambiguous to the data node.
 *
 * This is rather expensive and annoying to do once per row, but there's
 * little choice if we want to be sure values are transmitted accurately;
 * we can't leave the settings in place between rows for fear of affecting
 * user-visible computations.
 *
 * We use the equivalent of a function SET option to allow the settings to
 * persist only until the caller calls reset_transmission_modes().  If an
 * error is thrown in between, guc.c will take care of undoing the settings.
 *
 * The return value is the nestlevel that must be passed to
 * reset_transmission_modes() to undo things.
 */
int
set_transmission_modes(void)
{
	int nestlevel = NewGUCNestLevel();

	/*
	 * The values set here should match what pg_dump does.  See also
	 * configure_remote_session in connection.c.
	 */
	if (DateStyle != USE_ISO_DATES)
		(void) set_config_option("datestyle",
								 "ISO",
								 PGC_USERSET,
								 PGC_S_SESSION,
								 GUC_ACTION_SAVE,
								 true,
								 0,
								 false);
	if (IntervalStyle != INTSTYLE_POSTGRES)
		(void) set_config_option("intervalstyle",
								 "postgres",
								 PGC_USERSET,
								 PGC_S_SESSION,
								 GUC_ACTION_SAVE,
								 true,
								 0,
								 false);
	if (extra_float_digits < 3)
		(void) set_config_option("extra_float_digits",
								 "3",
								 PGC_USERSET,
								 PGC_S_SESSION,
								 GUC_ACTION_SAVE,
								 true,
								 0,
								 false);

	return nestlevel;
}

/*
 * Undo the effects of set_transmission_modes().
 */
void
reset_transmission_modes(int nestlevel)
{
	AtEOXact_GUC(true, nestlevel);
}
