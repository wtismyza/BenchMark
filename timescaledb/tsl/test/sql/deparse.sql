-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.

-- We compare information(\d+) about manually created tables with the ones that were recreated using deparse_table command.
-- There should be no diff.

\set ECHO errors
\ir include/deparse_func.sql

SELECT format('%s/results/deparse_create.out', :'TEST_OUTPUT_DIR') AS "CREATE_OUT",
       format('%s/results/deparse_recreate.out', :'TEST_OUTPUT_DIR') AS "RECREATE_OUT"
\gset
SELECT format('\! diff %s %s', :'CREATE_OUT', :'RECREATE_OUT') as "DIFF_CMD"
\gset

\ir include/deparse_create.sql

\o :CREATE_OUT
\d+ "public".*

\ir include/deparse_recreate.sql
\o :RECREATE_OUT
\d+ "public".*

\o

:DIFF_CMD

SELECT 'TABLE DEPARSE TEST DONE';

\set ECHO all
-- test drop_chunks function deparsing
SELECT * FROM tsl_test_deparse_drop_chunks('myschema.table10', '2019-01-01'::timestamptz, verbose => true);
SELECT * FROM tsl_test_deparse_drop_chunks('table1', newer_than => 12345);
SELECT * FROM tsl_test_deparse_drop_chunks('table1', older_than => interval '2 years', newer_than => '2015-01-01'::timestamp);

-- test generalized deparsing function
SELECT * FROM tsl_test_deparse_scalar_func(schema_name => 'Foo', table_name => 'bar', option => false, "time" => timestamp '2019-09-10 11:08', message => 'This is a test message.');

SELECT * FROM tsl_test_deparse_named_scalar_func(schema_name => 'Foo', table_name => 'bar', option => false, "time" => timestamp '2019-09-10 11:08', message => 'This is a test message.');

SELECT * FROM tsl_test_deparse_composite_func(schema_name => 'Foo', table_name => 'bar', option => false, "time" => timestamp '2019-09-10 11:08', message => 'This is a test message.');

-- test errors handling
\set ON_ERROR_STOP 0

CREATE TEMP TABLE fail_table1(x INT);

SELECT _timescaledb_internal.get_tabledef('fail_table1');

CREATE INDEX my_fail_table1_idx ON fail_table1 USING BTREE(x);

SELECT _timescaledb_internal.get_tabledef('my_fail_table1_idx');

SELECT _timescaledb_internal.get_tabledef('non_existing');

CREATE TABLE row_sec(i INT);
ALTER TABLE row_sec ENABLE ROW LEVEL SECURITY;
SELECT _timescaledb_internal.get_tabledef('row_sec');
