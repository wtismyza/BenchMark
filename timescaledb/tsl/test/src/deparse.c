/*
 * This file and its contents are licensed under the Timescale License.
 * Please see the included NOTICE for copyright information and
 * LICENSE-TIMESCALE for a copy of the license.
 */
#include <postgres.h>
#include <utils/builtins.h>
#include <access/htup.h>
#include <access/htup_details.h>
#include <catalog/pg_type.h>
#include <funcapi.h>

#include <utils.h>
#include <compat.h>
#include <extension.h>
#include "export.h"
#include "deparse.h"

TS_FUNCTION_INFO_V1(ts_test_get_tabledef);

Datum
ts_test_get_tabledef(PG_FUNCTION_ARGS)
{
	Oid relid = PG_GETARG_OID(0);
	const char *cmd = deparse_get_tabledef_commands_concat(relid);
	PG_RETURN_TEXT_P(cstring_to_text(cmd));
}

TS_FUNCTION_INFO_V1(ts_test_deparse_drop_chunks);

Datum
ts_test_deparse_drop_chunks(PG_FUNCTION_ARGS)
{
	FmgrInfo flinfo;
	FunctionCallInfo fcinfo2 = palloc(SizeForFunctionCallInfo(fcinfo->nargs));
	Oid argtypes[] = { REGCLASSOID, ANYOID, ANYOID, BOOLOID };
	Oid funcid = ts_get_function_oid("drop_chunks",
									 ts_extension_schema_name(),
									 sizeof(argtypes) / sizeof(*argtypes),
									 argtypes);
	const char *sql_cmd;
	int i;

	fmgr_info(funcid, &flinfo);
	InitFunctionCallInfoData(*fcinfo2,
							 &flinfo,
							 fcinfo->nargs,
							 fcinfo->fncollation,
							 fcinfo->context,
							 fcinfo->resultinfo);

	/* Copy over the arguments into the new function call data */
	for (i = 0; i < fcinfo->nargs; i++)
	{
		FC_ARG(fcinfo2, i) = FC_ARG(fcinfo, i);
		FC_NULL(fcinfo2, i) = FC_NULL(fcinfo, i);
	}

	/* Use the expression from this function so that the deparse function can
	 * result the data types of args with ANY type */
	fcinfo2->flinfo->fn_expr = fcinfo->flinfo->fn_expr;
	sql_cmd = deparse_func_call(fcinfo2);

	PG_RETURN_TEXT_P(cstring_to_text(sql_cmd));
}

TS_FUNCTION_INFO_V1(ts_test_deparse_func);

Datum
ts_test_deparse_func(PG_FUNCTION_ARGS)
{
	TupleDesc tupdesc;
	Oid resulttypeid;
	const char *deparsed = deparse_func_call(fcinfo);
	Datum retval = 0;

	elog(NOTICE, "Deparsed: %s", deparsed);

	switch (get_call_result_type(fcinfo, &resulttypeid, &tupdesc))
	{
		case TYPEFUNC_SCALAR:
			retval = BoolGetDatum(true);
			break;
		case TYPEFUNC_COMPOSITE:
		{
			Datum *values = palloc(tupdesc->natts * sizeof(Datum));
			bool *nulls = palloc(tupdesc->natts * sizeof(bool));
			HeapTuple tup;
			int i;

			for (i = 0; i < tupdesc->natts; i++)
				nulls[i] = true;

			tup = heap_form_tuple(tupdesc, values, nulls);
			pfree(values);
			pfree(nulls);
			retval = HeapTupleGetDatum(tup);
			break;
		}
		case TYPEFUNC_RECORD:
			/* indeterminate rowtype result */
		case TYPEFUNC_COMPOSITE_DOMAIN:
			/* domain over determinable rowtype result */
		case TYPEFUNC_OTHER:
			elog(ERROR, "unsupported result type for deparsing");
			break;
	}

	PG_RETURN_DATUM(retval);
}
