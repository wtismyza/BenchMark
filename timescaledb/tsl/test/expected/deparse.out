-- This file and its contents are licensed under the Timescale License.
-- Please see the included NOTICE for copyright information and
-- LICENSE-TIMESCALE for a copy of the license.
-- We compare information(\d+) about manually created tables with the ones that were recreated using deparse_table command.
-- There should be no diff.
\set ECHO errors
        ?column?         
-------------------------
 TABLE DEPARSE TEST DONE
(1 row)

-- test drop_chunks function deparsing
SELECT * FROM tsl_test_deparse_drop_chunks('myschema.table10', '2019-01-01'::timestamptz, verbose => true);
                                                                       tsl_test_deparse_drop_chunks                                                                        
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 SELECT * FROM public.drop_chunks(relation => 'myschema.table10',older_than => 'Tue Jan 01 00:00:00 2019 PST'::timestamp with time zone,newer_than => NULL,verbose => 't')
(1 row)

SELECT * FROM tsl_test_deparse_drop_chunks('table1', newer_than => 12345);
                                                  tsl_test_deparse_drop_chunks                                                  
--------------------------------------------------------------------------------------------------------------------------------
 SELECT * FROM public.drop_chunks(relation => 'public.table1',older_than => NULL,newer_than => '12345'::integer,verbose => 'f')
(1 row)

SELECT * FROM tsl_test_deparse_drop_chunks('table1', older_than => interval '2 years', newer_than => '2015-01-01'::timestamp);
                                                                              tsl_test_deparse_drop_chunks                                                                              
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 SELECT * FROM public.drop_chunks(relation => 'public.table1',older_than => '@ 2 years'::interval,newer_than => 'Thu Jan 01 00:00:00 2015'::timestamp without time zone,verbose => 'f')
(1 row)

-- test generalized deparsing function
SELECT * FROM tsl_test_deparse_scalar_func(schema_name => 'Foo', table_name => 'bar', option => false, "time" => timestamp '2019-09-10 11:08', message => 'This is a test message.');
NOTICE:  Deparsed: SELECT * FROM public.tsl_test_deparse_scalar_func(schema_name => 'Foo',table_name => 'bar',time => 'Tue Sep 10 11:08:00 2019 PDT',message => 'This is a test message.',not_set => 't',option => 'f')
 tsl_test_deparse_scalar_func 
------------------------------
 t
(1 row)

SELECT * FROM tsl_test_deparse_named_scalar_func(schema_name => 'Foo', table_name => 'bar', option => false, "time" => timestamp '2019-09-10 11:08', message => 'This is a test message.');
NOTICE:  Deparsed: SELECT option FROM public.tsl_test_deparse_named_scalar_func(schema_name => 'Foo',table_name => 'bar',time => 'Tue Sep 10 11:08:00 2019 PDT',message => 'This is a test message.',not_set => 't',option => 'f')
 option 
--------
 t
(1 row)

SELECT * FROM tsl_test_deparse_composite_func(schema_name => 'Foo', table_name => 'bar', option => false, "time" => timestamp '2019-09-10 11:08', message => 'This is a test message.');
NOTICE:  Deparsed: SELECT success,message FROM public.tsl_test_deparse_composite_func(schema_name => 'Foo',table_name => 'bar',time => 'Tue Sep 10 11:08:00 2019 PDT',message => 'This is a test message.',not_set => NULL,option => 'f')
 success | message 
---------+---------
         | 
(1 row)

-- test errors handling
\set ON_ERROR_STOP 0
CREATE TEMP TABLE fail_table1(x INT);
SELECT _timescaledb_internal.get_tabledef('fail_table1');
ERROR:  temporary table is not supported
CREATE INDEX my_fail_table1_idx ON fail_table1 USING BTREE(x);
SELECT _timescaledb_internal.get_tabledef('my_fail_table1_idx');
ERROR:  "my_fail_table1_idx" is an index
SELECT _timescaledb_internal.get_tabledef('non_existing');
ERROR:  relation "non_existing" does not exist at character 43
CREATE TABLE row_sec(i INT);
ALTER TABLE row_sec ENABLE ROW LEVEL SECURITY;
SELECT _timescaledb_internal.get_tabledef('row_sec');
ERROR:  row security is not supported
